casper/contracts/simple_casper.v.py

```diff
@@ -440,30 +443,39 @@ def logout(logout_msg: bytes <= 1024):
     end_dynasty: int128 = self.dynasty + self.DYNASTY_LOGOUT_DELAY
     assert self.validators[validator_index].end_dynasty > end_dynasty
 
-    # Set the end dynasty
     self.validators[validator_index].end_dynasty = end_dynasty
+    self.validators[validator_index].total_deposits_at_logout = self.total_curdyn_deposits_in_wei()
     self.dynasty_wei_delta[end_dynasty] -= self.validators[validator_index].deposit
 
     log.Logout(self.validators[validator_index].withdrawal_addr, validator_index, self.validators[validator_index].end_dynasty)
 
 
-
 # Withdraw deposited ether
 @public
 def withdraw(validator_index: int128):
     # Check that we can withdraw
     assert self.dynasty >= self.validators[validator_index].end_dynasty + 1
     end_epoch: int128 = self.dynasty_start_epoch[self.validators[validator_index].end_dynasty + 1]
-    assert self.current_epoch >= end_epoch + self.WITHDRAWAL_DELAY
+    withdrawal_epoch: int128 = end_epoch + self.WITHDRAWAL_DELAY
+    assert self.current_epoch >= withdrawal_epoch
+
     # Withdraw
-    withdraw_amount: int128(wei) = floor(self.validators[validator_index].deposit * self.deposit_scale_factor[end_epoch])
+    withdraw_amount: int128(wei)
+    if not self.validators[validator_index].is_slashed:
+        withdraw_amount = floor(self.validators[validator_index].deposit * self.deposit_scale_factor[end_epoch])
+    else:
+        recently_slashed: wei_value = self.total_slashed[withdrawal_epoch] - self.total_slashed[withdrawal_epoch - 2 * self.WITHDRAWAL_DELAY]
+        fraction_to_slash: decimal = recently_slashed * 3 / self.validators[validator_index].total_deposits_at_logout
+        fraction_to_withdraw: decimal = (1 - fraction_to_slash)
```

Comment:
1. would look nicer if we define the magic number 3 somewhere. 
2. if `total_deposit_at_logout = 100` at epoch 5, and then slashed 20 at epoch 5 and another 20 at epoch 6. we'll have a `fraction_to_slash = 40 * 3 /100 = 1.2` and thus a negative `fraction_to_withdraw`, negative `withdraw_amount`, and a failed send. so a mininum is required like ` fraction_to_withdraw  = min(1- fraction_to_slash, 0)`.

Reply:
Nice catch! I suppose it doesn't really matter if the withdrawal fails at this point, but we want the validator to get deleted and everything cleaned up. Actually, if they have nothing to withdraw, they probably will never call it... That's okay. 

Will do a `min` and only do the send if > 0.

Reply:
edit: it would need to be a `max(1- fraction_to_slash, 0)`